src/h_assets/cached_file.py

```diff
@@ -0,0 +1,75 @@
+import configparser
+import json
+from os.path import getmtime
+
+from pyramid.settings import aslist
+
+
+class CachedFile:  # pylint:disable=too-few-public-methods
+    """
+    Parses content from a file and caches the result.
+
+    _CachedFile reads a file at a given path and parses the content using a
+    provided loader.
+    """
+
+    path = None
+
+    def __init__(self, path, auto_reload):
+        """
+        Create the CachedFile object.
+
+        :param path: The path to the file to load
+        :param loader: A callable that parses content from a file
+        :param auto_reload: If True, the parsed content is discarded if the
+            mtime of the file changes.
+        """
+
+        self.path = path
+        self._auto_reload = auto_reload
+        self._last_modified_time = None
+        self._cached_content = None
+
+    def load(self):
+        """
+        Return the content of the file parsed with the loader.
+
+        The file is parsed when this is called for the first time and, if
+        auto-reload is enabled, when the mtime of the file changes.
+        """
+
+        current_mtime = getmtime(self.path)
+
+        if not self._cached_content or (
+            self._auto_reload and self._last_modified_time < current_mtime
+        ):
+            with open(self.path) as handle:
+                self._cached_content = self._load_handle(handle)
+
+            self._last_modified_time = current_mtime
+
+        return self._cached_content
+
+    @classmethod
+    def _load_handle(cls, handle):
+        return handle.read()
+
+
+class CachedJSONFile(CachedFile):  # pylint: disable=too-few-public-methods
+    @classmethod
+    def _load_handle(cls, handle):
+        return json.load(handle)
```

Comment:
To my mind the details of how to parse a manifest or bundle file belong with the `Environment`, since the `Environment` needs to know about the type/shape of the resulting data, whereas `CachedFile`'s responsibilities are determining whether the cached data exists and is fresh, and re-reading the bytes from the source if not.

Having convenience code which reads only generic rather than `Environment`-specific formats in this module would preserve the separation of responsibilities.

Reply:
I don't think there's anything environment specific about a JSON file and the line in the `Environment` makes it pretty clear what's going on:

```python
self.manifest_file = CachedJSONFile(manifest_path, auto_reload=auto_reload)
```

I think you're right about the INI file though. Initially I thought it was sort of false generality (we we only read one kind of file ever) as per the values discussion yesterday, but I'd give it a shot anyway. It actually revealed a different type of false generality.

In the previous formulation `load()` return the whole bundled section and called `aslist` on every part. Where it's used we only ever want one value from this. So actually I think this is a good split that lets you understand it better _and_ is actually more efficient. The only downside is that the bundle parsing and `aslist`-ing isn't behind the cache.

So the evolution of the things you'd have to understand to grok this part

* A more abstract `CacheFile`
* With a private function declared at the bottom of the file away from the `files()` method
* The `files()` and `urls()` methods including logic to read from it

Second draft:

* A concrete `CachedBundleFIle` which includes reading and parsing and some decoding locally within itself
* The `files()` method (with `urls()` calling it) keeping the look up in one place

This draft:

* A more abstract `CachedINIFile` which includes reading and parsing INI only all locally
* The `files()` method (with `urls()` calling it) doing all the format specific work

I think this is at a pretty good spot now.

Reply:
OK, this sounds better.